### ISSUE TITLE
YSQL: statement reads rows it has inserted. 

### ISSUE THREAD

**Issue by andrei-mart** [yugabyte/yugabyte-db#9436]
Given two one row tables:
```
yugabyte=# create table test1(id int primary key, value int);
CREATE TABLE
Time: 331.039 ms
yugabyte=# create table test2(id int primary key, value int);
CREATE TABLE
Time: 306.874 ms
yugabyte=# insert into test1 values(1, 1);
INSERT 0 1
Time: 50.970 ms
yugabyte=# insert into test2 values(2, 2);
INSERT 0 1
Time: 48.619 ms
yugabyte=# select id, value from test1 union all select id, value from test2;
 id | value 
----+-------
  1 |     1
  2 |     2
(2 rows)

Time: 68.971 ms
```
Following statement inserts 3 rows, while it is expected to insert two: 
```
yugabyte=# insert into test2 select id + 2, value from test1 union all select id + 2, value from test2;
INSERT 0 3
Time: 67.001 ms
```
Resulting data:
```
yugabyte=# select id, value from test2;
 id | value 
----+-------
  5 |     1
  4 |     2
  2 |     2
  3 |     1
(4 rows)

Time: 9.147 ms
```
What happens during execution of the plan
```
yugabyte=# explain insert into test2 select id + 2, value from test1 union all select id + 2, value from test2;
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Insert on test2  (cost=0.00..210.00 rows=2000 width=8)
   ->  Append  (cost=0.00..210.00 rows=2000 width=8)
         ->  Seq Scan on test1  (cost=0.00..100.00 rows=1000 width=8)
         ->  Seq Scan on test2 test2_1  (cost=0.00..100.00 rows=1000 width=8)
(4 rows)

Time: 3.909 ms
```
is that
- Append executes scan on test1
- One record is retrieved, passed up to Insert
- Insert adds (3, 1) tuple to test2
- Scan on test1 has no more rows, so Append executes scan on test2
- Two records are retrieved, one (2, 2) that should be visible, second (3, 1) that shouldn't 
- Insert respectively adds (4, 2) and (5, 1) into test2

Reproduced on a build from master
```
yugabyte=# select version();
                                                  version                                                   
------------------------------------------------------------------------------------------------------------
 PostgreSQL 11.2-YB-2.7.3.0-b0 on x86_64-pc-linux-gnu, compiled by gcc (Homebrew gcc 5.5.0_4) 5.5.0, 64-bit
(1 row)

Time: 24.183 ms
```

**Comment by deeps1991** [comment]
Assigning this to myself as a part of [#5310](https://github.com/yugabyte/yugabyte-db/issues/5310) has the exact same root cause as this one.

**Comment by deeps1991** [comment]
Closing, as issue is fixed in master and back ported to 2.6.
